File: latest/week1/workshop1_m4/W01_M04_demo.R
# ============================================================
# Workshop 1 — Module 4: Color Theory & Accessibility
# R Demonstration Script
# Data Visualization for Data Scientists — UNYT Tirana
# ============================================================

library(tidyverse)
# install.packages(c("viridis", "RColorBrewer", "colorspace", "patchwork"))
library(viridis)
library(RColorBrewer)
library(colorspace)
library(patchwork)

dir.create("output", showWarnings = FALSE)


# ── 1. Sequential Palette: Heatmap ─────────────────────────
set.seed(33)
heatmap_data <- expand_grid(
  day = c("Mon", "Tue", "Wed", "Thu", "Fri", "Sat"),
  week = paste0("W", 1:8)
) |>
  mutate(value = runif(n(), 10, 90))

p_heat <- ggplot(heatmap_data, aes(x = week, y = day, fill = value)) +
  geom_tile(color = "white", linewidth = 0.5) +
  scale_fill_viridis_c(option = "D", direction = 1) +
  theme_minimal(base_size = 10) +
  labs(title = "Sequential Palette: viridis",
       fill = "Value", x = NULL, y = NULL) +
  theme(plot.title = element_text(face = "bold"))

print(p_heat)
ggsave("output/heatmap_viridis.png", p_heat, width = 7, height = 4, dpi = 150)


# ── 2. Diverging Palette: Deviation from Mean ──────────────
set.seed(44)
df_div <- tibble(
  x = runif(80, 0, 10),
  y = runif(80, 0, 8),
  deviation = runif(80, -1, 1)
)

p_div <- ggplot(df_div, aes(x = x, y = y, color = deviation)) +
  geom_point(size = 3) +
  scale_color_gradient2(
    low = "#2166AC", mid = "white", high = "#B2182B",
    midpoint = 0, limits = c(-1, 1)
  ) +
  theme_minimal(base_size = 10) +
  labs(title = "Diverging Palette: RdBu Centred on Zero",
       color = "Deviation") +
  theme(plot.title = element_text(face = "bold"))

print(p_div)
ggsave("output/diverging_rdbu.png", p_div, width = 7, height = 5, dpi = 150)


# ── 3. Qualitative Palette: Categories ─────────────────────
set.seed(55)
df_qual <- tibble(
  category = sample(LETTERS[1:5], 100, replace = TRUE),
  x = rnorm(100),
  y = rnorm(100)
)

p_qual <- ggplot(df_qual, aes(x = x, y = y, color = category)) +
  geom_point(size = 2.5, alpha = 0.7) +
  scale_color_brewer(palette = "Set2") +
  theme_minimal(base_size = 10) +
  labs(title = "Qualitative Palette: ColorBrewer Set2",
       color = "Category") +
  theme(plot.title = element_text(face = "bold"))

print(p_qual)
ggsave("output/qualitative_set2.png", p_qual, width = 7, height = 5, dpi = 150)


# ── 4. Rainbow vs Viridis ─────────────────────────────────
set.seed(7)
df_rainbow <- tibble(
  x = runif(200, 0, 10),
  y = runif(200, 0, 10),
  z = sin(x) * cos(y) + rnorm(200, 0, 0.3)
)

p_rain <- ggplot(df_rainbow, aes(x = x, y = y, color = z)) +
  geom_point(size = 1.5) +
  scale_color_gradientn(colors = rainbow(20)) +
  theme_minimal(base_size = 9) +
  labs(title = "Rainbow (BAD)", color = "Value") +
  theme(plot.title = element_text(face = "bold", color = "#C62828"))

p_vir <- ggplot(df_rainbow, aes(x = x, y = y, color = z)) +
  geom_point(size = 1.5) +
  scale_color_viridis_c() +
  theme_minimal(base_size = 9) +
  labs(title = "Viridis (GOOD)", color = "Value") +
  theme(plot.title = element_text(face = "bold", color = "#1565C0"))

p_compare <- p_rain + p_vir +
  plot_annotation(title = "Never Use Rainbow")

print(p_compare)
ggsave("output/rainbow_vs_viridis.png", p_compare, width = 10, height = 4, dpi = 150)


# ── 5. CVD Simulation ─────────────────────────────────────
# The colorspace package simulates colour vision deficiency

p_original <- ggplot(df_qual, aes(x = x, y = y, color = category)) +
  geom_point(size = 2.5) +
  scale_color_manual(values = c(
    A = "#E53935", B = "#2E7D32", C = "#1565C0",
    D = "#FF9800", E = "#9C27B0"
  )) +
  theme_minimal(base_size = 9) +
  labs(title = "Normal Vision") +
  theme(plot.title = element_text(face = "bold"))

# Simulate deuteranopia
p_deutan <- deutan(p_original) +
  labs(title = "Deuteranopia Simulation") +
  theme(plot.title = element_text(face = "bold"))

p_cvd <- p_original + p_deutan +
  plot_annotation(title = "CVD Simulation with colorspace::deutan()")

print(p_cvd)
ggsave("output/cvd_simulation.png", p_cvd, width = 10, height = 4, dpi = 150)


# ── 6. Grey + Accent Strategy ─────────────────────────────
set.seed(55)
df_accent <- tibble(
  x = runif(100, 10, 100),
  y = 0.4 * x + rnorm(100, 0, 12),
  category = sample(c("A", "B", "C", "D"), 100, replace = TRUE),
  is_highlight = category == "D"
)

p_accent <- ggplot(df_accent, aes(x = x, y = y)) +
  # Background: grey
  geom_point(
    data = filter(df_accent, !is_highlight),
    color = "#BBBBBB", size = 1.5, alpha = 0.4
  ) +
  # Highlight: red accent
  geom_point(
    data = filter(df_accent, is_highlight),
    color = "#E53935", size = 3, alpha = 0.9
  ) +
  theme_minimal(base_size = 10) +
  labs(title = "Grey + Accent: Category D Highlighted",
       x = "Spend ($K)", y = "ROI (%)") +
  theme(plot.title = element_text(face = "bold"))

print(p_accent)
ggsave("output/grey_accent.png", p_accent, width = 7, height = 5, dpi = 150)


cat("\n── All M04 plots saved to output/ ──\n")

File: latest/week1/workshop1_m4/W01_M04_demo.py
"""
============================================================
Workshop 1 — Module 4: Color Theory & Accessibility
Python Demonstration Script
Data Visualization for Data Scientists — UNYT Tirana
============================================================
"""

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import os

os.makedirs("output", exist_ok=True)
plt.rcParams["figure.dpi"] = 150
np.random.seed(42)


# ── 1. Sequential Palette: Heatmap ─────────────────────────
data = np.random.uniform(10, 90, (6, 8))
days = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat"]
weeks = [f"W{i}" for i in range(1, 9)]

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(11, 4))

# Viridis (recommended)
im1 = ax1.imshow(data, cmap="viridis", aspect="auto")
ax1.set_xticks(range(8)); ax1.set_xticklabels(weeks, fontsize=7)
ax1.set_yticks(range(6)); ax1.set_yticklabels(days, fontsize=7)
plt.colorbar(im1, ax=ax1, shrink=0.8)
ax1.set_title("viridis (perceptually uniform)", fontsize=9, fontweight="bold")

# YlGnBu (ColorBrewer sequential)
im2 = ax2.imshow(data, cmap="YlGnBu", aspect="auto")
ax2.set_xticks(range(8)); ax2.set_xticklabels(weeks, fontsize=7)
ax2.set_yticks(range(6)); ax2.set_yticklabels(days, fontsize=7)
plt.colorbar(im2, ax=ax2, shrink=0.8)
ax2.set_title("YlGnBu (ColorBrewer)", fontsize=9, fontweight="bold")

fig.suptitle("Sequential Palettes for Quantitative Data", fontsize=11, fontweight="bold")
plt.tight_layout()
plt.savefig("output/heatmap_sequential_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 2. Diverging Palette ──────────────────────────────────
np.random.seed(44)
x = np.random.uniform(0, 10, 80)
y = np.random.uniform(0, 8, 80)
z = np.random.uniform(-1, 1, 80)

fig, ax = plt.subplots(figsize=(7, 5))
sc = ax.scatter(x, y, c=z, cmap="RdBu_r", s=60, vmin=-1, vmax=1,
                edgecolors="white", linewidth=0.5)
plt.colorbar(sc, ax=ax, label="Deviation from Mean")
ax.set_title("Diverging Palette: RdBu (centred on zero)", fontsize=10, fontweight="bold")
ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
plt.tight_layout()
plt.savefig("output/diverging_rdbu_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 3. Rainbow vs Viridis ─────────────────────────────────
np.random.seed(7)
x = np.random.uniform(0, 10, 200)
y = np.random.uniform(0, 10, 200)
z = np.sin(x) * np.cos(y) + np.random.normal(0, 0.3, 200)

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 4))

ax1.scatter(x, y, c=z, cmap="rainbow", s=20, edgecolors="none")
ax1.set_title("Rainbow (BAD)\nFalse contours, CVD-hostile",
              fontsize=9, fontweight="bold", color="#C62828")

ax2.scatter(x, y, c=z, cmap="viridis", s=20, edgecolors="none")
ax2.set_title("Viridis (GOOD)\nPerceptually uniform, CVD-safe",
              fontsize=9, fontweight="bold", color="#1565C0")

fig.suptitle("Never Use Rainbow", fontsize=12, fontweight="bold")
plt.tight_layout()
plt.savefig("output/rainbow_vs_viridis_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 4. CVD-Safe Alternative: cividis ──────────────────────
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 4))

ax1.scatter(x, y, c=z, cmap="viridis", s=20, edgecolors="none")
ax1.set_title("viridis", fontsize=9, fontweight="bold")

ax2.scatter(x, y, c=z, cmap="cividis", s=20, edgecolors="none")
ax2.set_title("cividis (optimised for CVD)", fontsize=9, fontweight="bold")

fig.suptitle("CVD-Safe Alternatives", fontsize=11, fontweight="bold")
plt.tight_layout()
plt.savefig("output/cividis_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 5. Grey + Accent Strategy ─────────────────────────────
np.random.seed(55)
n = 100
xs = np.random.uniform(10, 100, n)
ys = 0.4 * xs + np.random.normal(0, 12, n)
cats = np.random.choice(["A", "B", "C", "D"], n)

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 4))

# All coloured (confusing)
palette = {"A": "#1f77b4", "B": "#ff7f0e", "C": "#2ca02c", "D": "#d62728"}
for c in ["A", "B", "C", "D"]:
    mask = cats == c
    ax1.scatter(xs[mask], ys[mask], c=palette[c], s=30, alpha=0.6, label=c)
ax1.legend(fontsize=7, title="Cat")
ax1.set_title("All Coloured (cluttered)", fontsize=9, fontweight="bold", color="#C62828")
ax1.spines["top"].set_visible(False)
ax1.spines["right"].set_visible(False)

# Grey + accent
mask_hl = cats == "D"
ax2.scatter(xs[~mask_hl], ys[~mask_hl], c="#BBBBBB", s=20, alpha=0.3, label="Other")
ax2.scatter(xs[mask_hl], ys[mask_hl], c="#E53935", s=50, alpha=0.9,
            edgecolors="white", linewidth=0.5, zorder=5, label="Category D")
ax2.legend(fontsize=7)
ax2.set_title("Grey + Accent (focused)", fontsize=9, fontweight="bold", color="#1565C0")
ax2.spines["top"].set_visible(False)
ax2.spines["right"].set_visible(False)

fig.suptitle("Strategic Colour: Grey + Accent", fontsize=12, fontweight="bold")
plt.tight_layout()
plt.savefig("output/grey_accent_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 6. Qualitative Palettes Compared ─────────────────────
cats_list = ["Cat A", "Cat B", "Cat C", "Cat D", "Cat E"]
vals = [34, 45, 28, 52, 38]

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 4))

# Saturated rainbow (bad)
rainbow = ["#FF0000", "#00FF00", "#0000FF", "#FFFF00", "#FF00FF"]
ax1.bar(cats_list, vals, color=rainbow, edgecolor="black", linewidth=1)
ax1.set_title("Saturated Rainbow (BAD)", fontsize=9, fontweight="bold", color="#C62828")
ax1.set_facecolor("#F5F5F5")

# Muted Tableau 10 (good)
muted = ["#4e79a7", "#f28e2b", "#e15759", "#76b7b2", "#59a14f"]
ax2.bar(cats_list, vals, color=muted, edgecolor="white", linewidth=1)
ax2.set_title("Tableau 10 Muted (GOOD)", fontsize=9, fontweight="bold", color="#1565C0")
ax2.spines["top"].set_visible(False)
ax2.spines["right"].set_visible(False)

fig.suptitle("Qualitative Palettes: Muted > Saturated", fontsize=11, fontweight="bold")
plt.tight_layout()
plt.savefig("output/qualitative_comparison_py.png", dpi=150, bbox_inches="tight")
plt.close()


print("── All M04 plots saved to output/ ──")
